### PR TITLE
Feature/telemetry

### PR DESCRIPTION
**Pull request by chinmoy12c**

### CODE REVIEW

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 1]
```diff
@@ -1,19 +1,48 @@
...
+    constructor(private config: Record<string, any>) { }
 
-    static doesConsumeEvent(eventName: string): Boolean {
+    static doesConsumeEvent(eventName: string): boolean {
         return false;
     }
 
     getName(): string {
         throw new Error("Method not implemented.");
     }
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:15]: You need to implement this to return a unique name like `Telemetry`.

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 2]
```diff
@@ -1,19 +1,48 @@
+import axios from 'axios';
 import { ISideEffect } from "../common/sideEffect.interface";
 import { SideEffectData } from "../common/sideEffect.types";
 
 export class TelemetrySideEffect implements ISideEffect {
 
-    constructor(config: Record<string, any>) { }
+    constructor(private config: Record<string, any>) { }
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:7]:
```suggestion
    constructor(config: Record<string, any>) { }
```

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 3]
```diff
@@ -1,19 +1,48 @@
...
 
-    static doesConsumeEvent(eventName: string): Boolean {
+    static doesConsumeEvent(eventName: string): boolean {
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:9]:
```suggestion
    static doesConsumeEvent(eventName: string): Boolean {
```

File: packages/xmessage/src/types.ts [hunk 4]
```diff
@@ -70,9 +70,17 @@ export enum DeviceType {
...
+  transformerId: string,
+  eventName: string,
+  timestamp: string,
+  eventData: Record<string, any>,
+}
+
 export class Transformer {
   id?: string;
   metaData?: Record<string, any>;
+  sideEffectData?: SideEffectData;
```
Comment (chinmoy12c) [inline packages/xmessage/src/types.ts:83]: Revert this, we already have this definition in `common`

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 5]
```diff
@@ -1,19 +1,48 @@
...
+            return true;
+        } catch (error) {
+            console.error("Error occurred during telemetry:", error);
+            return false;
+        }
+    }
+
+    private extractHostFromEventData(eventData: Record<string, any>): string {
+        return eventData?.transformer?.metaData?.host || "";
+    }
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:38]: Assuming `host` to be a part of `metaData` is wrong. You should be accepting it from `config[getName()].host`

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 6]
```diff
@@ -1,19 +1,48 @@
...
+        try {
+            const eventData = sideEffectData.eventData?.payload;
+            const host = this.extractHostFromEventData(eventData);
+
+            if (!eventData || !host) {
+                console.error("Event data or host not found.");
+                return false;
+            }
+
+            await this.sendEventDataToTemetry(eventData, host);
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:27]: Are we sending the entire `payload` to telemetry, has this been discussed in the spec with @dhruv-1001? I think there are specific types that you can use in the columns.

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 7]
```diff
@@ -1,19 +1,96 @@
...
+
+    private createEventData(sideEffectData: SideEffectData): ConstructedEventData {
+        const subEventData: SubEventData = {
+            botId: "botId",
+            userId: sideEffectData.eventData.to.userID,
+            conversationId: "conversationId",
+            orgId: sideEffectData.eventData?.orgId || '',
+            messageId: sideEffectData.eventData?.messageId || '',
+            error: sideEffectData.error,
+            phoneNumber: "phonenumber",
```
Comment (RyanWalker277) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:67]: Hey @chinmoy12c, I'm not sure how are we going to extract the botId, conversationId and phonenumber when the transformer gets executed? Wanted to know your thoughts on this.

Comment (chinmoy12c) [reply packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:67]: According to XMessage spec:
- You would get `botId` in `eventData.app`
- conversationId in `messageId.Id`
- You would "not" get phoneNumber since a message may not always have a phone number, although you would get the associated userid(which can be the phone number, depends on the channel being used) in `to.userId`

Comment (RyanWalker277) [reply packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:67]: Understood. I don't think we should include the phone number then as we are already passing in userId.

File: packages/sideEffects/src/common/sideEffect.types.ts [hunk 8]
```diff
@@ -1,8 +1,12 @@
 import { XMessage } from "@samagra-x/xmessage";
+import { version } from '../../node_modules/@types/uuid/index.d';
 
 export type SideEffectData = {
     transformerId: string,
     eventName: string,
     timestamp: number,
+    version:string;
+    environment:string,
+    phonenumber:;
```
Comment (chinmoy12c) [inline packages/sideEffects/src/common/sideEffect.types.ts:10]: We cannot modify this, please revert this change

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 9]
```diff
@@ -1,19 +1,96 @@
...
+    timeTaken?: number;
+}
 
-    constructor(config: Record<string, any>) { }
+export class TelemetrySideEffect implements ISideEffect {
+    constructor(private readonly config: Record<string, any>) { }
 
-    static doesConsumeEvent(eventName: string): Boolean {
+    static doesConsumeEvent(eventName: string): boolean {
         return false;
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:33]: You need to tell whether an event with `eventName` is being consumed by this side effect. Returning false makes your side effect "not work" by default. Check [here](https://github.com/BharatSahAIyak/orchestrator/pull/17/files#) for getting both start and end event names.

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 10]
```diff
@@ -1,19 +1,96 @@
...
+        const eventData: ConstructedEventData = {
+            version: sideEffectData.version,
+            eventType: "single",
+            event: sideEffectData.eventName,
+            subEvent: subEventData,
+            timestamp: Date.now(),
+            actorId: sideEffectData.transformerId,
+            actorType: "Transformer",
+            environment: sideEffectData.environment,
+        };
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:79]: You would need to handle `start` and `end` events differently. You might want to pass `totalTime` in `end` event.

File: packages/sideEffects/src/common/sideEffect.interface.ts [hunk 11]
```diff
@@ -2,5 +2,5 @@ import { SideEffectData } from "./sideEffect.types";
 
 export interface ISideEffect {
     getName(): string;
-    execute(sideEffectData: SideEffectData): Promise<Boolean>;
+    execute(sideEffectData: SideEffectData[]): Promise<Boolean>;
```
Comment (chinmoy12c) [inline packages/sideEffects/src/common/sideEffect.interface.ts:5]:
```suggestion
    execute(sideEffectData: SideEffectData): Promise<Boolean>;
```

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 12]
```diff
@@ -1,19 +1,110 @@
...
     }
 
     getName(): string {
-        throw new Error("Method not implemented.");
+        return "Telemetry";
+    }
+
+    async execute(sideEffectData: SideEffectData[]): Promise<boolean> {
+        try {
+            const eventDataArray: ConstructedEventData[] = sideEffectData.map(data => this.createEventData(data));
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:46]: `sideEffectData` will not be array.

File: packages/sideEffects/src/telemetry/telemetry.sideEffect.ts [hunk 13]
```diff
@@ -1,19 +1,110 @@
...
+
+        const eventData: ConstructedEventData = {
+            generator: "Transformer",
+            version: "0.0.1", // Placeholder
+            timestamp: new Date().toISOString(),
+            actorId: sideEffectData.transformerId,
+            actorType: "Transformer",
+            sessionId: "123412341234", // Placeholder
+            deviceId: "123412341234", // Placeholder
+            env: "prod", // Placeholder
```
Comment (chinmoy12c) [inline packages/sideEffects/src/telemetry/telemetry.sideEffect.ts:79]: Why are these placeholder.
